core/src/main/scala/kafka/log/ProducerIdMapping.scala

```diff
@@ -54,48 +56,106 @@ private[log] class ProducerAppendInfo(val pid: Long, initialEntry: ProducerIdEnt
   private var lastSeq = initialEntry.lastSeq
   private var lastOffset = initialEntry.lastOffset
   private var maxTimestamp = initialEntry.timestamp
+  private var currentTxnFirstOffset = initialEntry.currentTxnFirstOffset
+  private val startedTxns = ListBuffer.empty[OngoingTxn]
```

Comment:
I realized that we need all started transactions in order to ensure that the LSO is updated correctly. We take all the started transactions from an append, add them to a sorted collection, and then remove the completed transactions in order of completion. This ensures that the LSO is always correct.